### PR TITLE
[IMP] point_of_sale: backport of a2319b4 + fix and complete the task

### PR DESCRIPTION
**Pull request by wtaferner**
There was no reason why this was not merged in v10 as it works
seamlessly and the commit in master was missing out invoiced orders
reconciliation, commercial partner and eventually comparing apples with
oranges :-)

**Description of the issue/feature this PR addresses:**
As mentioned in a2319b4 the renconciliation was not done automatically although we have all facts needed during closing a session normally.

**Current behavior before PR:**
Reconciliation tasks are left behind and mostly forgotten by the users and will just end up in a huge amount of unreconciled entries in the database.

**Desired behavior after PR is merged:**
Reconciliation is done automatically if possible during the closing of a PoS session

@nim-odoo 
Thank you for the work in master and I hope you like my additional work on this. With this little change in the code the point of sale is much more fun even in v10 :smiley: 

@mart-e @jorenvo 
May I ask you for a review. As mentioned in the commit message I would appreciate if this could be merged in stable.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by mart-e** [comment]
Hello Wolfgang,

I believe we already discussed this on #15005 with @nim-odoo 
It is a change of behaviour and not a trivial change.

**Comment by wtaferner** [comment]
You are right Martin that it was discussed before, but after backporting and examining it in detail I believe it is a trivial change and not changing anything which would not be done manually. 

Maybe you could reconsider your position? The plus here is that I fixed issues of the master commit as well which are future bugfixes ;-)

I really would appreciate and I am sure a lot of others would support this too.

**Comment by wtaferner** [comment]
Pleeeaseee :smiley: 

At least give me a more concrete use case why you think it is such a change in behavior that it is not worth the "risk".

**Comment by wpichler** [comment]
Hi all,

@mart-e i also think that it is a change of behaviour - but it is a necessary change of behaviour - because the current behaviour is wrong. So it is a bug fix in my opinion.

Please merge it into stable

**Comment by pedrobaeza** [comment]
@wtaferner this can be easily extracted in a community module that copy/paste the same method and overwrites _confirm_orders, calling super and calling the method, so I think that's the way to go to have it on v10.

**Comment by mart-e** [comment]
@nim-odoo @qdp-odoo opinion?

**Comment by qdp-odoo** [comment]
I don't have any problem with that...

**Comment by gustavovalverde** [comment]
This is a great FIX. Good Job @wtaferner

**Comment by wtaferner** [comment]
@nim-odoo 
I think your opinion is missing or maybe @mart-e is waiting for you to merge? :-) But I guess, Quentin should have removed all insecurities about the decision to merge, right?

My customers are super happy with the fix and it works smoothly :-)

Thank you all for the support and thx @qdp-odoo for taking the time to review.

**Comment by wtaferner** [comment]
Ok, just figured out that we might need this in case a user is not also a user of at least purchase/user.

Not a big deal, but better safe than sorry :-)

**Comment by mart-e** [comment]
> Or should we go for sudo?

You do not want to use sudo to reconcile entries in a multicompany environment 😉 
Actually I am surprised you can close a session without the accounting rights.

**Comment by nim-odoo** [comment]
Still not a big fan of changing the behavior in 10.0, but if QDP agrees... Note that it could even be backported to 9.0.

Maybe a nitpicking comment but I would:
- cherry-pick a2319b4
- add the modifications in an additional commit
I think that would make the forward-port smoother.

**Comment by wtaferner** [comment]
@nim-odoo 
Honor to whom it belongs, so I did what you have asked for.
About v9, I really think it would be too much, but feel free to take my PR and apply it on v9, do test it as good as possible and forward port to v10 asap :-)

@mart-e 
Right, but if we start from orders of a session in company A it won't happen to get lines from company B, right? Otherwise it would be awkward...

**Comment by wpichler** [comment]
one suggestion here - why don't you put a config option on pos.config model - auto_reconciliation - which does defaults to False for v10, v9 and v8. If it is False - then just proceed with the old behaviour - if it is set to true then do the auto reconcile.
So it will be completly no change in behaviour

File: addons/point_of_sale/models/pos_order.py [hunk 1]
```diff
@@ -334,15 +334,16 @@ def insert_data(data_type, values):
 
     def _reconcile_payments(self):
         for order in self:
-            aml = order.statement_ids.mapped('journal_entry_ids') | order.account_move.line_ids
-            aml = aml.filtered(lambda r: not r.reconciled and r.account_id.internal_type == 'receivable' and r.partner_id == order.partner_id)
+            aml = order.statement_ids.mapped('journal_entry_ids').mapped('line_ids') | order.account_move.line_ids | order.invoice_id.move_id.line_ids
```
Comment (nim-odoo) [inline addons/point_of_sale/models/pos_order.py]: I was wondering how I could make that mistake, but I realized that `journal_entry_ids` are `account.move.line` from saas-14, but `account.move` in v10.0.

Comment (wtaferner) [inline addons/point_of_sale/models/pos_order.py]: Makes sense, so not your fault :-) I thought I was checking on master, but obviously you checked again, I was too fast, but yeah, it was not my focus during backport :-)

**Comment by mart-e** [comment]
> one suggestion here - why don't you put a config option on pos.config model - auto_reconciliation 

No sorry @wpichler , adding an option is not the way to go. It makes the software more complex, more code to maintain,... And anyway adding an option means adding a field which is not possible in stable.

@wtaferner actually, adding the security part is problematic. It means that, for existing db, you must upgrade the point of sale to be able to close a session. This is not acceptable for stable version.
@qdp-odoo is it safe to use `sudo` on a `reconcile` call? No user related operation?

**Comment by qdp-odoo** [comment]
@mart-e where do you see a sudo on reconcile? There is no reason to have any...

**Comment by wtaferner** [comment]
@mart-e 
Well, as said it is nothing which would really block as user security rules are administration stuff and easy to resolve, but honestly I really would go for the ```sudo``` anyway as it is used for the other [accounting parts](https://github.com/odoo/odoo/blob/10.0/addons/point_of_sale/models/pos_order.py#L150) of session closing too and as said, it won't mess up between companies as it gets all lines related to the orders of a explicit company session of their related moves, invoices, statements and tries to reconcile between them.

The sudo would actually be done on ```orders.sudo()._reconcile_payments()```

**Comment by mart-e** [comment]
Ok for sudo (and remove the csv part then).
(and when we asked why we are so reluctant to backport to stable, it is to prevent this kind of regression with lack of access rights.)

**Comment by wtaferner** [comment]
@mart-e 
Here we go, and sure, even access rights are for some users an issue, but at least in this PR we could remove this obstacle again.

**Comment by wtaferner** [comment]
@nim-odoo @mart-e 
Are you waiting each other that one of you is merging or is still anything missing?

**Comment by mart-e** [comment]
Thanks for the update and your patience, it has finally been merged at https://github.com/odoo/odoo/commit/fe70f0711f2a198aa84a5bdbe1e1db703cb52831

**Pull request merged**